### ISSUE TITLE
Add option for concurrency limiting per subgraph

### ISSUE THREAD

**Issue by spurrkins** [apollographql/router#2973]
In NodeJS, we're able to limit concurrency per subgraph by passing in a custom agent with maxSockets set. This restricts how many connections can be open with every target, preventing concurrency from blowing up. This is particularly useful as defense against subgraph failure modes that result in extremely elevated latency. If the subgraph receives enough volume that requests outpace timeouts, this could result in FD exhaustion, OOMKill on the gateway/router, and make subgraph recovery more difficult. My organization has experienced two brief outages on the gateway due to situations like this in the past, where one broken subgraph caused OOMKills that hurt use cases completely unrelated to the broken subgraph.

I think an option in the Traffic Shaping section to configure a max socket count would help mitigate this risk, as I believe the router is just as vulnerable to this as most NodeJS implementations. I'd also like to see this avoid what I view as a pitfall of the NodeJS Http Agent - [once it reaches the socket count, it enqueues requests](https://github.com/nodejs/node/blob/965ea9d586a49a42328f32aa339dc661e7ebf26d/lib/_http_agent.js#L295-L308), which also will result in an OOMKill. A more reasonable behavior would be to fast fail a request to a subgraph if there's no socket capacity to service it.

**Comment by goto-bus-stop** [comment]
I don't think this is specifically planned currently, however OOM as a result of this should be mostly resolved in router 2.x and will be further addressed by the upcoming introduction of circuit breaking.
